### PR TITLE
Add tracker measures (combined train\validation\test loss and accuracy) in output towards TensorBoard

### PR DESCRIPTION
**Pull request by carlogrisetti**
This is the implementation of my own feature request #350 
"works fine on my machine" as usual, and since it's based on the same value calculated for the CLI output, treating it just as a number and adding it to the same step, it has no impact on any of the calculations. It's using the same file_writer, appending a "generated" summary containing only that one value to the corresponding step.

### CODE REVIEW

**Comment by w4nderlust** [comment]
lgtm thank you Carlo. Will test it out soon and. then merge

**Comment by w4nderlust** [comment]
Tested, looks fine, but actually why not adding all the values in the tracker instead of only the validation loss? :) Let me know if you are willing to add them, otherwise I will do it myself after merging this.

**Comment by carlogrisetti** [comment]
Oh well... That was the only useful value to me at the time (and still now it is). You mean validation accuracy, test loss, and test accuracy, right?
Since there's a different granularity on training metrics acquired via the usual training loop and the ones calculated at the end of the epoch, do you want training statistics also?

I wouldn't think of an use case for any "live" metric other than validation loss/accuracy, but you surely know better :)

**Comment by carlogrisetti** [comment]
Ok so... I've added everything (I just thought... it's free, and you just can not look at that if you don't need those).
The only "open point" I see is with the formatting we want to give to that.

- All values under one "evaluation" group (as it is now with this latest commit)
- Separate "evaluation_training" \ "evaluation_validation" \ "evaluation_test" groups?

I can't make my mind up on how to display those... also since this is an ubiquitous metric that will always be present in TensorBoard, it might be useful to force its ordering to be the first one (ie: the old awful "!" to force it to be alphabetically sorted as first).

Also, it might be useful to move train_reg_mean_loss under the same metrics group, since it's almost the same measure (or at least it means almost the same thing) as the evaluation\training_loss, with a different granularity.

**Comment by w4nderlust** [comment]
Hey @carlogrisetti , the tests don;t pass and I undesrand why: you assume that there's an accuracy measure, but in some cases there is not. Instead of assuming what measures are there you can do something as simple as cycling through the ones available in the progress tracker, something like:
```
def add_tensorboard_summary(stats, prefix, train_writer, step):
    summaries = []
    for output_f in stats:
        for metric in output_f:
            metric_tag = "evaluation/{}_{}_{}".fomart(prefix, output_f, metric)
            metric_val = test_stats[output_f][metric][-1]
            summaries.append(
                tf.compat.v1.Summary.Value(tag=metric_tag, simple_value=metric_val)
             )
     test_summary = tf.compat.v1.Summary(value=summaries)
     train_writer.add_summary(test_summary, step)

add_tensorboard_summary(progress_tracker.train_stats, "train", train_writer, progress_tracker.steps - 1)
add_tensorboard_summary(progress_tracker.vali_stats, "validation", train_writer, progress_tracker.steps - 1)
add_tensorboard_summary(progress_tracker.test_stats, "test", train_writer, progress_tracker.steps - 1)
```
This will dinamically adapt to whatever is in the tracker.

Regarding "evaluation/" I think  it's fine, we can change later the older summaries to have a "train_batch" prefix, so that it's clear that evaluation is for each epoch, train batch for each batch.

**Comment by carlogrisetti** [comment]
... couldn't have done it better... and in fact I didn't :)

That makes totally sense... I just reformatted and declared that as a classmethod since it has all external parameters passed to it and uses no class objects implicitly. I hope in turn it makes sense to you. I am targeting only the combined measures, since they're the only ones having a "global" meaning in regards to the training.

On the naming convention maybe we could go this way:
- `evaluation_epoch` (or `evaluation_by_epoch`, or even `Evaluation (by epoch)`) under which to move everything i've added here)`
- `evaluation_batch` (or `evaluation_by_batch`, or even `Evaluation (by batch)`) under which to move train_reg_mean_loss only, since all the others are grouped by feature)
I wouldn't call them batch_evaluation and epoch_evaluation (prefixing the granularity), just for the sake of alphabetical ordering and thus graphical output in TensorBoard.

If that's ok, I could also edit the train_reg_mean_loss to prefix with the `evaluation_batch/` tag (or whatever you choose to be the best fit) and push it alltogether.

**Comment by w4nderlust** [comment]
I'm not sure if one could do nested groups like "evaluation/feature_name/measure" and "batch/feature_name/measure", but in case we probably wnat to do that.

**Comment by carlogrisetti** [comment]
It would be the ideal arrangement, but unfortunately you can't do that and are limited to one level of grouping, so batch/epoch "tag" and feature name should be in some way combined.
We could also prepend "batch_" to any batch-level measure and "epoch_" to the tracker ones I've added (as epoch_combined)

**Comment by w4nderlust** [comment]
> It would be the ideal arrangement, but unfortunately you can't do that and are limited to one level of grouping, so batch/epoch "tag" and feature name should be in some way combined.
> We could also prepend "batch_" to any batch-level measure and "epoch_" to the tracker ones I've added (as epoch_combined)

Got it, yes that makes sense!

**Comment by carlogrisetti** [comment]
I'm having some trouble finding where to inject the "batch_" prepend in the single measure tag for all the other features, since it seems this is handled autonomously by TF, and I don't want to mess too much with the inner workings of Ludwig (I don't really have the expertise to do that). Any help you can provide on that?

However, I think this could already be worthy of a merge at the current state. If all the already existing measures get a "batch_" prepend, this could be also worthy of a changelog mention in the next release, not to get confusing when comparing old training results with a new one.

**Comment by w4nderlust** [comment]
You can search for "summary" and see where they are, but I can already tell you there is one in model.py (train_reg_mean_loss) and the other ones are in the output feature objects of each datatype. For instance, for numerical, you have to look in `features/numerical_feature.py` and summaries definitions are in `NumericalOutputFeature.build_output()`. The same is true for all other feature types.
If it's too much work don't worry, I can do it myself after we merge your PR.

**Comment by carlogrisetti** [comment]
I already patched the `train_reg_mean_loss` as `epoch_combined/train_reg_mean_loss` in the last commit.

I understood the other values are built in each feature, but couldn't understand how they are given a name. Maybe, now knowing I was looking in the right spot, I'll try to have a look at it tomorrow and otherwise will defer to you for fixing that last thing.

Have a good one

**Comment by carlogrisetti** [comment]
I never got around to do this, and I don't know when I'll have time to look deeper into that.
Up to you to choose to wait, fix yourself, or merge as is

**Comment by w4nderlust** [comment]
No worries, will merge and work on the naming convention.
